File: LeetCode/Easy/Increasing Order Search Tree/C++/search_tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        vector<TreeNode*> in;
        inorder(root, in);
        for (int i=1; i<in.size(); i++) {
            in[i-1]->left = nullptr;
            in[i-1]->right = in[i];
        }
        in[in.size()-1]->left = nullptr;
        in[in.size()-1]->right = nullptr;
        return in[0];
    }

    void inorder(TreeNode* root, vector<TreeNode*>& in) {
        if (root == nullptr) return;
        inorder(root->left, in);
        in.push_back(root);
        inorder(root->right, in);
    }
};

File: LeetCode/Easy/Increasing Order Search Tree/Java/searchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<TreeNode> in;
    public TreeNode increasingBST(TreeNode root) {
        in = new ArrayList<>();
        inorder(root);
        for (int i=1; i<in.size(); i++) {
            in.get(i-1).left = null;
            in.get(i-1).right = in.get(i);
        }
        in.get(in.size()-1).left = null;
        in.get(in.size()-1).right = null;
        return in.get(0);
    }

    private void inorder(TreeNode root) {
        if (root == null) return;
        inorder(root.left);
        in.add(root);
        inorder(root.right);
    }
}
